### ISSUE TITLE
Poor scalability on Multi-Core CPU systems (OpenBLAS to blame)

### ISSUE THREAD

**Issue by Technologov** [LeelaChessZero/lc0#652]
looking at Leela again, I can tell you that it's basic algoritm is scalable, if it can run on a GPU.  Non-parallel algo cannot.

But it's implementation on a CPU is a disaster -- OpenBLAS backend to blame.

**** I call to find a replacement for OpenBLAS **** for Multi-Core CPU systems.

Take a look:
Totally new Dual Xeon Linux server.

BigFenix: Dual Intel(R) Xeon(R) CPU E5-2697A v4 @ 2.60GHz (32 core; 64-thread) + Ubuntu 18.04 Linux x64 (Broadwell) 
(OpenBLAS)  Engine : Leela 0.20.0 + Neural Network ID 32400 (20/256).
~/leela-0.20/lc0 benchmark -t 32

thread- 1: Nodes/second :  42
thread- 4: Nodes/second : 142
thread- 8: Nodes/second : 220
thread-16: Nodes/second : 270 (!)
thread-32: Nodes/second : 222 
thread-64: Nodes/second : 115 (WOOT !!! OMG !!! Such a slowdown on hyper-threaded CPU !)

For the sake of comparison of what a well-written algorithm looks like on a CPU: (stockfish-9)

BigFenix: Dual Intel(R) Xeon(R) CPU E5-2697A v4 @ 2.60GHz (32 core; 64-thread) + Linux x64
stockfish-9-linux/Linux/stockfish-9-bmi2 bench 16 64
thread- 1: Nodes/second    :  2204312
thread- 2: Nodes/second    :  4216857
thread- 4: Nodes/second    :  8221277
thread- 8: Nodes/second    : 16288761
thread-16: Nodes/second    : 32518175
thread-32: Nodes/second    : 56162013  (25.5x times faster ! Yay !)
thread-64: Nodes/second    : 71519468  (32.4x times faster ! Yay !)

stockfish-9 scales very well on servers and workstations, up to 64 processors.

Another example of poorly written multi-Core algorithm: Stockfish again ! Version 6. (stockfish-6)

BigFenix: Dual Intel(R) Xeon(R) CPU E5-2697A v4 @ 2.60GHz (32 core; 64-thread) + Linux x64
stockfish-6-linux/Linux/stockfish_6_x64_bmi2 bench 16 $CPU_CORES

thread- 1: Nodes/second    :  2342722
thread- 2: Nodes/second    :  3738296
thread- 4: Nodes/second    :  4828941
thread- 8: Nodes/second    :  3747961 (Regression !)
thread-16: Nodes/second    :  2706391
thread-32: Nodes/second    :  1671791 (Disaster ! Performance falls under 1-core CPU !)
thread-64: Nodes/second    :  770190 (Terrible.)

Basically stockfish-6 cannot scale above 4 or 6 cores, and it's performance drops like a stone.

**Comment by gcp** [comment]
OpenBLAS threading is a huge nuisance, but I'm not sure you have evidence here that's the problem? Is lc0 tuned for more than 32 threads?

Upstream Leela has Eigen as an alternate backend, but it's a bit more annoying for release packages because it only supports the CPU architecture it's built on. It's not a dynamic lib like OpenBLAS.

**Comment by gcp** [comment]
>Chiming in as I was looking into threading strategies for BLAS.

The BLAS shouldn't be threaded at all, as was explicitly pointed out in the post before yours (it's disabled!). The engine should be capable of generating enough threads with independent work.

When I complained about OpenBLAS threading, I'm referring to it having compile-time limitations on the maximum number of threads even when you're not using the threading at all (MAX_CPUS etc). Eigen doesn't have this problem. But this has got nothing to do with threading the BLAS computation itself, which something like lc0 would want to strongly avoid.

**Comment by Naphthalin** [comment]
Given the performane on CCC/TCEC CPU hardware, I think this issue can be considered to be solved.
